File: client/mock-server/game/targets.ts
import { WS } from '../serverTexts';
import type { PieceKind, TargetSpec } from './catalog';
import { EffectError, parsePlacement, parseSquare, pieceColor, pieceName, relativeRank, type Color } from './fen';
import { KIND_WALL } from './squares';
import type { Tracker } from './tracker';

/**
 * Porting di `effects/targets.go`: ogni bersaglio rispetta il suo `TargetSpec`, nello stesso ordine. I bersagli sono
 * caselle distinte e il re non è mai un bersaglio, salvo un `own_piece` che lo elenca in `pieces`.
 */

export function validateTargets(fen: string, tracker: Tracker, specs: readonly TargetSpec[], targets: readonly string[], caster: Color): void {
  const grid = parsePlacement(fen);
  const seen = new Set<string>();
  specs.forEach((spec, index) => {
    const square = targets[index];
    if (square === undefined) return;
    const reject = (reason: string, message: string) => new EffectError(WS.target(index, reason, square, message));

    let row: number;
    let col: number;
    try {
      [row, col] = parseSquare(square);
    } catch {
      throw reject('off_board', `casella non valida: ${JSON.stringify(square)}`);
    }
    if (seen.has(square)) throw reject('duplicate', `la casella ${square} è già stata scelta`);
    seen.add(square);
    const piece = grid[row]?.[col] ?? null;

    if (spec.type === 'square') {
      if (spec.empty_square === true && piece !== null) throw reject('not_empty', `la casella ${square} non è vuota`);
      // Una casa col muro non è vuota: niente muri impilati, niente pezzi evocati o teletrasportati lì.
      if (spec.empty_square === true && tracker.hasSquareEffect(square, KIND_WALL)) throw reject('wall', `la casella ${square} ha un muro`);
    } else {
      if (piece === null) throw reject('no_piece', `nessun pezzo in ${square}`);
      const own = pieceColor(piece) === caster;
      if (own !== (spec.type === 'own_piece')) throw reject('wrong_owner', `il pezzo in ${square} è del colore sbagliato`);
      const kindReason = pieceKindRefusal(spec, pieceName(piece) as PieceKind);
      if (kindReason !== null) throw reject(kindReason, `il pezzo in ${square} non è un bersaglio ammesso`);
      if (spec.require_effect !== undefined && !tracker.hasEffect(square, spec.require_effect)) {
        throw reject('missing_effect', `il pezzo in ${square} non ha l'effetto ${spec.require_effect}`);
      }
    }

    const previous = targets[index - 1];
    if (spec.max_distance !== undefined && spec.max_distance > 0 && previous !== undefined) {
      const d = chebyshev(previous, square);
      if (d > spec.max_distance) throw reject('too_far', `${square} è a distanza ${d}, massimo ${spec.max_distance}`);
    }
    const rank = relativeRank(square, caster);
    if (spec.own_ranks !== undefined && spec.own_ranks.length > 0 && !spec.own_ranks.includes(rank)) {
      throw reject('rank', `la traversa di ${square} non è ammessa`);
    }
    if (spec.min_rank !== undefined && spec.min_rank > 0 && rank < spec.min_rank) {
      throw reject('rank', `${square} è prima della traversa ${spec.min_rank}`);
    }
  });
}

function pieceKindRefusal(spec: TargetSpec, kind: PieceKind): string | null {
  if (kind === 'king') return spec.type === 'own_piece' && spec.pieces?.includes('king') === true ? null : 'king';
  if (spec.pieces !== undefined && spec.pieces.length > 0 && !spec.pieces.includes(kind)) return 'piece_kind';
  return null;
}

function chebyshev(a: string, b: string): number {
  try {
    const [ar, ac] = parseSquare(a);
    const [br, bc] = parseSquare(b);
    return Math.max(Math.abs(ar - br), Math.abs(ac - bc));
  } catch {
    return 99;
  }
}

File: server/internal/effects/targets.go
package effects

import (
	"chess-server/internal/gameerr"
	"chess-server/internal/spells"
)

// Motivi di rifiuto di un bersaglio, in details.reason di invalid_target.
const (
	ReasonOffBoard      = "off_board"      // casella inesistente
	ReasonDuplicate     = "duplicate"      // stessa casella scelta due volte
	ReasonNotEmpty      = "not_empty"      // serviva una casa vuota
	ReasonNoPiece       = "no_piece"       // serviva un pezzo
	ReasonWrongOwner    = "wrong_owner"    // pezzo del colore sbagliato
	ReasonKing          = "king"           // il re non è un bersaglio valido
	ReasonPieceKind     = "piece_kind"     // tipo di pezzo non ammesso
	ReasonMissingEffect = "missing_effect" // manca lo stato richiesto (es. freeze)
	ReasonTooFar        = "too_far"        // oltre MaxDistance dal bersaglio precedente
	ReasonRank          = "rank"           // traversa non ammessa
)

// ValidateTargets verifica che i bersagli rispettino i TargetSpec della magia,
// nello stesso ordine. È pura: legge la FEN e gli stati del Tracker senza
// modificarli. Il numero di bersagli lo verifica già match.CastSpell.
//
// Regole globali, valide per ogni magia: i bersagli sono caselle distinte e il
// re non è mai un bersaglio, salvo un TargetOwnPiece che lo elenca
// esplicitamente in Pieces.
func ValidateTargets(fen string, t *Tracker, specs []spells.TargetSpec, targets []string, caster Color) error {
	grid, err := parsePlacement(fen)
	if err != nil {
		return gameerr.Newf(gameerr.Internal, "FEN non valida: %v", err)
	}
	seen := make(map[string]bool, len(targets))
	for i, spec := range specs {
		if i >= len(targets) {
			break
		}
		square := targets[i]
		reject := func(reason, format string, args ...interface{}) error {
			return gameerr.Newf(gameerr.InvalidTarget, format, args...).
				With("index", i).With("reason", reason).With("square", square)
		}

		row, col, err := parseSquare(square)
		if err != nil {
			return reject(ReasonOffBoard, "casella non valida: %q", square)
		}
		if seen[square] {
			return reject(ReasonDuplicate, "la casella %s è già stata scelta", square)
		}
		seen[square] = true
		p := grid[row][col]

		switch spec.Type {
		case spells.TargetSquare:
			if spec.EmptySquare && p != 0 {
				return reject(ReasonNotEmpty, "la casella %s non è vuota", square)
			}
			// Una casa col muro non è vuota: niente muri impilati, niente pezzi evocati o teletrasportati lì.
			if spec.EmptySquare && t.HasSquareEffect(square, KindWall) {
				return reject(ReasonWall, "la casella %s ha un muro", square)
			}
		case spells.TargetOwnPiece, spells.TargetEnemyPiece:
			if p == 0 {
				return reject(ReasonNoPiece, "nessun pezzo in %s", square)
			}
			own := pieceColor(p) == caster
			if own != (spec.Type == spells.TargetOwnPiece) {
				return reject(ReasonWrongOwner, "il pezzo in %s è del colore sbagliato", square)
			}
			if reason := checkPieceKind(spec, p); reason != "" {
				return reject(reason, "il pezzo in %s non è un bersaglio ammesso", square)
			}
			if spec.RequireEffect != "" && !t.HasEffect(square, spec.RequireEffect) {
				return reject(ReasonMissingEffect, "il pezzo in %s non ha l'effetto %s", square, spec.RequireEffect)
			}
		default:
			return gameerr.Newf(gameerr.Internal, "tipo di bersaglio non supportato: %s", spec.Type)
		}

		if spec.MaxDistance > 0 && i > 0 {
			if d := chebyshev(targets[i-1], square); d > spec.MaxDistance {
				return reject(ReasonTooFar, "%s è a distanza %d, massimo %d", square, d, spec.MaxDistance)
			}
		}
		rank := RelativeRank(square, caster)
		if len(spec.OwnRanks) > 0 && !containsInt(spec.OwnRanks, rank) {
			return reject(ReasonRank, "la traversa di %s non è ammessa", square)
		}
		if spec.MinRank > 0 && rank < spec.MinRank {
			return reject(ReasonRank, "%s è prima della traversa %d", square, spec.MinRank)
		}
	}
	return nil
}

// checkPieceKind restituisce il motivo di rifiuto per il tipo di pezzo, o "".
func checkPieceKind(spec spells.TargetSpec, p byte) string {
	kind := spells.PieceKind(pieceName(p))
	if kind == spells.King {
		if spec.Type == spells.TargetOwnPiece && containsKind(spec.Pieces, spells.King) {
			return ""
		}
		return ReasonKing
	}
	if len(spec.Pieces) > 0 && !containsKind(spec.Pieces, kind) {
		return ReasonPieceKind
	}
	return ""
}

// chebyshev è la distanza "da re" fra due caselle valide.
func chebyshev(a, b string) int {
	ar, ac, errA := parseSquare(a)
	br, bc, errB := parseSquare(b)
	if errA != nil || errB != nil {
		return 99
	}
	return max(abs(ar-br), abs(ac-bc))
}

func abs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}

func containsInt(list []int, v int) bool {
	for _, x := range list {
		if x == v {
			return true
		}
	}
	return false
}

func containsKind(list []spells.PieceKind, v spells.PieceKind) bool {
	for _, x := range list {
		if x == v {
			return true
		}
	}
	return false
}
